src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -85,6 +93,15 @@ public Http2Connection(Http2ConnectionContext context)
             _hpackDecoder = new HPackDecoder((int)_serverSettings.HeaderTableSize);
         }
 
+        // internal for testing
+        internal Dictionary<ConnectionState, ManualResetEventSlim> StateTransitionEvents { get; set; }
```

Comment:
👍 For the action over the dictionary. Better yet, just test behavior instead of verifying arbitrary internal state transitions happen.

I also almost always recommend using TaskCompletionSource with TaskCreationOptions.RunContinuationsAsynchronously instead of ManualResetEvents since the latter encourages blocking using Wait() which can lead to real problems on some CI agents with low core counts. It's better to use TCS.Task.TimeoutAfter()/DefaultTimeout(.)